### ISSUE TITLE
ask about chapter 4.4.2 가중치 규제 추가 

### ISSUE THREAD

**Issue by moonk1212** [sju-coml/SAI-2020-Team-I#16]
4.4.2 가중치 규제 추가

상황: 1조다음은2조에서 스터디 하던 중 가중치 규제 추가 라는 부분에 대해서 얘기 나누던중가중치 규제 추가가 네트워크 복잡도에 제한을 두어 가중치가 작은 값을 가지도록 강제한다는 것은 알게 됨

문제: 네트워크 복잡도와 가중치 규제가 정확히 어떤 연관이 있는지 모르겠음
L1규제와 L2 규제에 대한 부분이 정확히 이해가 안되는 상황이 발생함
그래서 가중치 규제 추가 부분이 어떻게 과대적합을 막는지 정확히 이해가 안됨

**Comment by gon2gon2** [comment]
** L1 규제는 일부 가중치 파라미터를 0으로 만듭니다. L2 규제는 가중치 파라미터를 제한하지만 완전히 0으로 만들지는 않습니다. 
 "간단한 모델"은 모델 파라미터의 분포를 봤을 때 엔트로피(entropy)가 작은 모델입니다(또는 앞 절에서 보았듯이 적은 파라미터를 가진 모델입니다). 따라서 과대적합을 완화시키는 일반적인 방법은 가중치가 작은 값을 가지도록 네트워크의 복잡도에 제약을 가하는 것입니다 **
출처: https://www.tensorflow.org/tutorials/keras/overfit_and_underfit?hl=ko
가중치 파라미터를 아예 0으로 만들거나 작게 만든다 -> 가질 수 있는 값을(값의 분포)를 적게 해서 모델을 단순하게 만든다.
이게 맞을까요? (질문)

**Comment by moonk1212** [comment]
네네 맞는거 같습니다. 저희팀 질문에 답변 주셔서 감사합니다.@gon2gon2

**Comment by ProtossDragoon** [comment]
@moonk1212  @gon2gon2 

좋은 질의응답이네요! 두분 다 고생하셨습니다.

저는 조금 더 경험적인 답변을 해 드릴게요. 우리는 loss 를 통해 모델의 가중치가 적절히 적절한 값으로 수렴하도록 만듭니다. loss 는 보통 네트워크의 끝부분에서 결과가 연산이 되는 구조이지요. 이때, loss 에 L1 loss 나 L2 loss 를 추가하면, 모델이 가중치를 가지고 있다는 그 자체만으로도 혼을 내게 됩니다. 양수이든, 음수이든, 절댓값이 0이 아니라는 자체만으로도 혼내게 됩니다.

하나 더, 생각해볼만한 것은 "모델이 담는 정보량" 과  "가중치가 0인 것" 이 무슨 연관이 있는가? 에 대한 것입니다. 수학적으로 이해하는 방법도 존재하겠지만, 간단히 생각해 봅시다. 가중치가 존재하지 않는다는 것은, "특정 요소를 고려하지 않겠다" 와 동치라는 점에 동의하나요? 이때 0이 많다는 것은, "특정 요소를 고려하지 않고도 무엇인가를 추론해낼 수 있는" 모델을 요구하는 것이 됩니다. L1 Loss 와 L2 Loss 는 동일하게, 특정 가중치가 매우 큰 경우 엄벌을 내리게 되는데, 그것은 "특정 가중치가 크다는 것은, 특정 부분에 지나치게 의존하고 있다는 것" 이라고 해석할 수 있기 때문입니다.

<hr>

L1 Loss 와 L2 Loss 의 차이를 생각해 봅시다. 단순히 그래프를 떠올려 보면, L1 Loss 는 V자형 그래프가 만들어지므로, 0을 제외한 모든 점에서의 미분계수가 동일합니다. 0에 가깝든, 0에서 멀든 동일하게 혼내겠다는 말이 됩니다. 반면, L2 Loss 의 그래프를 생각해 본다면, U 자형 그래프가 만들어집니다. 이 경우, 0 에 가까워질수록, "덜 혼내겠다" 는 말이 됩니다. 따라서, L1 loss 는, 0에 가까워도 0이 되도록 혼내 버리기 때문에, L2 loss 보다 더 쉽게 0 으로 수렴할 수 있게 되고, 이러한 맥락에서 모델 압축에 사용한다는 말이 있기도 합니다.

<hr>

아래 내용은 이해하기 어려울 수 있습니다.

예를 들어, object detection 문제에서, 바운딩 박스를 적절히 찾도록 하는 loss 가 다음과 같다고 칩시다.

 L_final = L_bbox_regression + L_class_crossentropy

L 은 Loss 의 약자입니다. 이 L_final 은,  바운딩 박스를 정확히 "regression" 하도록 설계된 loss 와 class 를 잘 예측하도록 하는 "classification loss" 가 합쳐져 있네요.  이때, L2 loss 를 추가하게 되면

 L_final = L_bbox_regression + L_class_crossentropy + L_l2loss

이런 식으로 Loss 가 정의 되고, 역전파가 일어나게 됩니다.

**Comment by moonk1212** [comment]
이해하기 쉽게 설명해주셔서 감사합니다! 
@ProtossDragoon
